src/model/idtypes.rs

```diff
@@ -0,0 +1,300 @@
+use crate::model::Type;
+use serde::{Deserialize, Serialize};
+use std::borrow::Borrow;
+use std::marker::PhantomData;
+use strum::Display;
+use thiserror::Error;
+
+// This is a sealed trait pattern implementation, it stops external code from implementing the `IdType` trait.
+// The `Sealed` trait must be in a private mod, so external code can not see and implement it.
+// See also: https://rust-lang.github.io/api-guidelines/future-proofing.html
+mod private {
+    pub trait Sealed {}
+}
+
+pub trait IdType: private::Sealed {
+    const TYPE: Type;
+}
+pub trait PlayableIdType: IdType {}
+pub trait PlayContextIdType: IdType {}
+
+impl IdType for Artist {
+    const TYPE: Type = Type::Artist;
+}
+impl PlayContextIdType for Artist {}
+impl IdType for Album {
+    const TYPE: Type = Type::Album;
+}
+impl PlayContextIdType for Album {}
+impl IdType for Track {
+    const TYPE: Type = Type::Track;
+}
+impl PlayableIdType for Track {}
+impl IdType for Playlist {
+    const TYPE: Type = Type::Playlist;
+}
+impl PlayContextIdType for Playlist {}
+impl IdType for User {
+    const TYPE: Type = Type::User;
+}
+impl IdType for Show {
+    const TYPE: Type = Type::Show;
+}
+impl PlayContextIdType for Show {}
+impl IdType for Episode {
+    const TYPE: Type = Type::Episode;
+}
+impl PlayableIdType for Episode {}
+
+pub type ArtistId = Id<Artist>;
+pub type AlbumId = Id<Album>;
+pub type TrackId = Id<Track>;
+pub type PlaylistId = Id<Playlist>;
+pub type UserId = Id<User>;
+pub type ShowId = Id<Show>;
+pub type EpisodeId = Id<Episode>;
+
+pub type ArtistIdBuf = IdBuf<Artist>;
+pub type AlbumIdBuf = IdBuf<Album>;
+pub type TrackIdBuf = IdBuf<Track>;
+pub type PlaylistIdBuf = IdBuf<Playlist>;
+pub type UserIdBuf = IdBuf<User>;
+pub type ShowIdBuf = IdBuf<Show>;
+pub type EpisodeIdBuf = IdBuf<Episode>;
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Artist {}
+impl private::Sealed for Artist {}
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Album {}
+impl private::Sealed for Album {}
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Track {}
+impl private::Sealed for Track {}
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Playlist {}
+impl private::Sealed for Playlist {}
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum User {}
+impl private::Sealed for User {}
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Show {}
+impl private::Sealed for Show {}
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Episode {}
+impl private::Sealed for Episode {}
```

Comment:
These `impl IdType for xxType`, `xxxId`, `xxxIdBuf`, `enum xxx` statements are a bit verbose, is there a way to define them cleanly? for example, use a `Macro` to do so?

Reply:
Sure, I can use either a macro or a [sealed](https://crates.io/crates/sealed) crate here.